Longformer

ibeltagy opened the pull request:
## Adding the Longformer model
Paper: https://arxiv.org/abs/2004.05150
The code is mostly copied from https://github.com/allenai/longformer
Addressing issue: https://github.com/huggingface/transformers/issues/3783

**TODOs**
- [x] Add unit tests
- [x] Address PR reviews
- [x] Upload model to HF instead of AllenAI s3 buckets. 

**TODOs for other PRs**
- Add `LongformerForSequenceClassification`, `LongformerForMultipleChoice`, `LongformerForTokenClassification`, `LongformerForQuestionAnswering` with automatic setting of global attention based on the task.
- Add `LongformerEncoderDecoder`
- Add an example to show how to convert existing pretrained models into their long version.

stefan-it commented on src/transformers/tokenization_longformer.py:
```diff
@@ -0,0 +1,43 @@
+merges_url = "https://s3.amazonaws.com/models.huggingface.co/bert/roberta-large-merges.txt"
+_all_longformer_models = ["longformer-base-4096", "longformer-large-4096"]
+
+VOCAB_FILES_NAMES = {"vocab_file": "sentence.bpe.model"}
```
I think this should be something like:

```python
VOCAB_FILES_NAMES = {
    "vocab_file": "vocab.json",
    "merges_file": "merges.txt",
}
```

ibeltagy replied:
ok, I will check. I copied it from `tokenization_bart.py` https://github.com/huggingface/transformers/blob/master/src/transformers/tokenization_bart.py#L30

bratao commented:
This is awesome @ibeltagy . I´m very excited about longformer.

In the original implementation a good speedup was thanks to the personalized kernel in tvm. In this pull, this speedup is lost, or it was adapted somehow?

iliaschalkidis commented:
Thanx @ibeltagy! 

Is it possible for you or someone else who is familiar with TF2 to submit also a PR for TF2? I tried to migrate your code in the past but I couldn’t migrate the new attention mechanism written for TVM, so I moved on. Your work alongside Reformer has special interest for people like me and my colleagues, who are usually doing research with long documents (>=1000words). 

Maybe @thomwolf and @julien-c know the right person...

patrickvonplaten commented on src/transformers/modeling_longformer.py:
```diff
@@ -0,0 +1,667 @@
+
+# TODO: upload model and config files
+LONGFORMER_PRETRAINED_MODEL_ARCHIVE_MAP = {
+    "longformer-base-4096": "https://ai2-s2-research.s3-us-west-2.amazonaws.com/longformer/longformer-base-4096/pytorch_model.bin",
```
I can take care of uploading the model to the HF model page :-)

patrickvonplaten replied:
Actually, I think you guys already have an organization `allenai` - so I think you can upload the models to `allenai/longformer-base-4096` and `allenai/longformer-large-4096`. Since we have "organizations" now for models we can directly link it to them :-)

patrickvonplaten commented on src/transformers/modeling_longformer.py:
```diff
@@ -0,0 +1,667 @@
+        x_padded = x_padded.view(*x_padded.size()[:-2], x_padded.size(-1), x_padded.size(-2))
+        return x_padded
+
+    def _skew2(self, x, padding_value):
```
does the `padding_value` correspond to the `config.pad_token_id` here?

patrickvonplaten replied:
Not sure if that's the case. But if `padding_value` always corresponds to `config.pad_token_id` we could save it once in each class and access it from the class instead of using it as an argument to the functions.

ibeltagy replied:
`padding_value` is a large negative value to set attention to zero after softmax. I will clarify this.

patrickvonplaten commented on src/transformers/modeling_longformer.py:
```diff
@@ -0,0 +1,667 @@
+            diagonals_list.append(diagonal_mask)
+        mask = torch.stack(diagonals_list, dim=-1)
+        mask = mask[None, :, None, :]
+        ending_mask = mask.flip(dims=(1, 3)).bool().to(device)
```
@sshleifer @thomwolf - I think `bool()` break backward compatibility, no? I think we still use `torch.uint8` instead at the moment. So i think we should change `.bool()` to `.to(torch.uint8)` here

patrickvonplaten commented on src/transformers/modeling_longformer.py:
```diff
@@ -0,0 +1,667 @@
+        mask = torch.stack(diagonals_list, dim=-1)
+        mask = mask[None, :, None, :]
+        ending_mask = mask.flip(dims=(1, 3)).bool().to(device)
+        return affected_seq_len, mask.bool().to(device), ending_mask
```
same  for `bool()`

patrickvonplaten commented on src/transformers/modeling_longformer.py:
```diff
@@ -0,0 +1,667 @@
+        ending_mask = ending_mask[:, -seq_len:].expand(ending_input.size())
+        ending_input.masked_fill_(ending_mask, -float('inf'))
+
+    def _sliding_chunks_matmul_qk(self, q: torch.Tensor, k: torch.Tensor, w: int, padding_value: float):
```
Awesome to use type hinting here :-)

patrickvonplaten replied:
padding_value is `float` not `int` ?

ibeltagy replied:
yes, usually float('-inf')

sshleifer replied:
can it be a module level or class level constant?

patrickvonplaten commented on src/transformers/modeling_longformer.py:
```diff
@@ -0,0 +1,667 @@
+        This implementation splits the input into overlapping chunks of size 2w (e.g. 512 for pretrained Longformer)
+        with an overlap of size w'''
+        bsz, seqlen, num_heads, head_dim = q.size()
+        assert seqlen % (w * 2) == 0
```
Maybe add some assert statements here?

patrickvonplaten commented on src/transformers/modeling_longformer.py:
```diff
@@ -0,0 +1,667 @@
+        '''Matrix multiplicatio of query x key tensors using with a sliding window attention pattern.
+        This implementation splits the input into overlapping chunks of size 2w (e.g. 512 for pretrained Longformer)
+        with an overlap of size w'''
+        bsz, seqlen, num_heads, head_dim = q.size()
```
(nit) I like it better to write out `batch_size` here since this is the usual way we write it in the library.

patrickvonplaten commented on src/transformers/modeling_longformer.py:
```diff
@@ -0,0 +1,667 @@
+            float_mask = remove_from_windowed_attention_mask.type_as(q).masked_fill(remove_from_windowed_attention_mask, -10000.0)
+            ones = float_mask.new_ones(size=float_mask.size())  # tensor of ones
+            # diagonal mask with zeros everywhere and -inf inplace of padding
+            d_mask = self._sliding_chunks_matmul_qk(ones, float_mask, self.one_sided_attention_window_size, padding_value=0)
```
This hardcoded padding value here makes me a bit uneasy - related to the comments above...

patrickvonplaten commented on src/transformers/modeling_longformer.py:
```diff
@@ -0,0 +1,667 @@
+        attn_weights = attn_weights_float.type_as(attn_weights)
+        attn_probs = F.dropout(attn_weights_float.type_as(attn_weights), p=self.dropout, training=self.training)
+        v = v.view(seq_len, bsz, self.num_heads, self.head_dim).transpose(0, 1)
+        attn = 0
```
I think this `attn=0` can be a bit confusing for the reader because it will be transformed to a tensor in the lines that follow. I would prefer to set `attn = None` here and then replace the `attn += ...` line with an if statement `if attn is not None: ... else ...`.

patrickvonplaten commented on src/transformers/modeling_longformer.py:
```diff
@@ -0,0 +1,667 @@
+        attn += self._sliding_chunks_matmul_pv(attn_probs, v, self.one_sided_attention_window_size)
+
+        attn = attn.type_as(hidden_states)
+        assert list(attn.size()) == [bsz, seq_len, self.num_heads, self.head_dim]
```
Maybe add an error message ?

sshleifer replied:
you can also assert tuple equality without casting.

patrickvonplaten commented on src/transformers/modeling_longformer.py:
```diff
@@ -0,0 +1,667 @@
+            key_padding_mask = None
+
+        hidden_states = hidden_states.transpose(0, 1)
+        seq_len, bsz, embed_dim = hidden_states.size()
```
(nit) sometimes it's called `seqlen`, sometimes `seq_len`. Not at all important, but might be nice to decide on one naming for consistency :-)

patrickvonplaten commented on src/transformers/modeling_longformer.py:
```diff
@@ -0,0 +1,667 @@
+                f'Expected {config.num_hidden_layers}, given {len(config.attention_window)}'
+            )
+
+        if config.attention_mode == 'n2':
```
Do we need that `if config.attention_mode == `n2` at all? Is there a difference to Bert when using this setup `config.attention_mode == 'n2'`?

ibeltagy replied:
A better name is welcomed, 'bert', 'full', 'standard'?

While with "n2" this model is equivalent to RoBERTa, our pretrained model has a longer position embedding. It can be useful for speed/memory comparisons.

sshleifer replied:
maybe it should just be a bool e.g `use_longformer_attention=True`

patrickvonplaten commented on src/transformers/modeling_longformer.py:
```diff
@@ -0,0 +1,667 @@
+                # pad with position_id = pad_token_id as in modeling_roberta.RobertaEmbeddings
+                position_ids = F.pad(position_ids, (0, padding_len), value=pad_token_id)
+            if inputs_embeds is not None:
+                inputs_embeds = F.pad(inputs_embeds, (0, 0, 0, padding_len), value=0.0)  # pad with zero embeddings
```
I think it's usually safer to concatenate the embeddings of the `pad_token_id` to the `input_embeds` in order to pad the `input_embeds`, so that we don't have to hard code the value 0.0 here.

patrickvonplaten commented on src/transformers/modeling_longformer.py:
```diff
@@ -0,0 +1,667 @@
+            pad_token_id=self.config.pad_token_id)
+
+        # embed
+        output = super().forward(
```
I guess I would prefer to copy past the `forward` function of Roberta into the model here, but I guess ok for now! What do you think @thomwolf @LysandreJik ?

ibeltagy replied:
I am curious why you think this is better. I was assuming it is the other way around, that less duplicate code is better.

patrickvonplaten replied:
I guess that's actually a very fundamental point of implementing Longformer so here is my opinion: 

For me, `RobertaModel` == `BertModel` because it's literally the exact the same architecture except for the word embeddings which are completely overwritten in `RobertaModel` - so it's ok for me that `RobertaModel` inherits the complete forward function from Bert. `Longformer` on the other hand uses a lot of the things from `RobertaModel`: Feed Forward layers, the word embeddings (in it's expanded version), ..., but is essentially a different architecture and needs a slightly different forward function. It feels a bit to me that we try to wrap certain Longformer functionality around Roberta functionality and when it doesn't work overwrite very specific parts of `Roberta` logic with `Longformer` logic (longformer attention layer), whereas for me `Longformer` is essentially a different model. 

For me the downsides of using Roberta/Bert code as much as possible, but having a completely different attention layers are the following:

1. _Flexibility to add code / change code_: 
a) The attention layers are usually the most important part of a model so that functions at model level insert parameters for the attention layer. This being said, adding new features for Longformer won't really be possible without changing Bert code. Let's say we decide to also add a auto-regressive longformer version `LonformerWithLMHead` (not masked Longformer model) and want to add a caching mechanism to store past key-value states as is done in `gpt2`, `t5` ... (and soon `reformer`), we will anyways need a separate `forward` function. 

b) Longformer forward is completely dependant on Bert forward:
Every change to Bert's forward function affects Longformer (and Roberta). For Roberta, I think this is alright because it's the exact same architecture (same attention layer), so a change in `attention_masks`, `encoder_hidden_states`, etc. are fine, but could lead to problems in Longformer because it has a different attention layer. I'm a bit scared that it might make us write the Longformer code in a way that would fit Bert's forward function, but doesn't show the logic of longformer anymore in a clear way

2. _Readability_:
If a user wants to track what happens in Longformer when calling model(input_ids), he will have to jump from longformer's part forward to Roberta's non-existing forward to finally Bert. Errors will be very hard to track and understand.

The only upside for me is less duplicated code, but I very much prefer readability over duplicated code :-)

IMO, to keep maximum flexibility for Longformer I would go as far as to add a `LongformerLayer` class with its own `forward` function **and** a `LongformerAttention` class with its own forward function. I would then fill the layers in those classes either with "Bert components" (like `BertOutput`, `BertIntermediate`, ...) or the `LongformerSelfAttention`. Then we would not have to do this very specific layer injection at `LongformerModel` level. 

Also, people will soon ask about a encoder-decoder version for `Longformer` and I think at this point we probably should (or have to) rewrite the class to be able to handle cross longformer attention layers and longformer self attention layers well. I think the similarity between Bert/Roberta and Longformer is not big enough for complete model inheritance. 

Sorry for that long text :D 
But I guess I'm also quite biased, so very keen to discuss this :-) 


@LysandreJik, @thomwolf, @sshleifer  - what do you think?

thomwolf replied:
I agree with @patrickvonplaten.

One of the specificities of `transformers` (quite different from other libraries indeed) is to have each model code almost fully contained in a single file. This allows in particular people to easily extend the models.

We should do the same here and have the full code of Bert copied and modified here (at the expense of code duplication but it's fine for us), I think.

sshleifer replied:
I sort of agree with @ibeltagy 
 
(1) I think the PR that makes `LongformerEncoderDecoder` should copy paste if needed, but it is not required in this PR. I would also imagine that that PR will use this as the encoder and then write a new `LongformerDecoder` whose `forward` method supports incremental decoding.

(2) Most of the papers I've read describe their contributions on top of Bert, so it's nice to have the code reflect that abstraction.

(3) I think this is extremely readable. It's very clear that an external `forward` method is being called.

ibeltagy replied:
I second the arguments of @sshleifer. 

Here's a different suggestion for how to implement Longformer; only keep `LongformerSelfAttention` (no `LongformerModel`) and add an if statement inside various models to switch between `BertSelfAttention` and `LongformerSelfAttention` using a configuration. 
Pros: can convert existing models into their long version
Cons: a large code change
I don't think we should do this now though, just keep it in mind for future improvements.

patrickvonplaten replied:
Hmm, don't really agree with the arguments @sshleifer :D 

1) - I don't really think we should  add much code for the Encoder-Decoder version besides the `cross_attention_layer`. We should use the EncoderDecoderModel from https://github.com/huggingface/transformers/blob/7defc6670fa76e857109e1b99f3e919da8d11f42/src/transformers/modeling_encoder_decoder.py#L29 to create the encoder and decoder from  `LongformerModel` as it's already done in Bert. There should not be a `LongformerDecoder` IMO.

2) - I think the code should show which `Roberta/Bert` components were used and which components are new. I don't think we need a high level abstraction for this.

3) - Most people have installed the library with pip. If they run `Longformer.forward` and see `Bert` specific error messages, I could see this leading to confusion. 

I agree though that this (big) change does not have to be done in this PR though, but can be done at a later stage.

ibeltagy replied:
Will keep this to a separate PR.

patrickvonplaten commented:
@ibeltagy - Looks awesome the PR :-) I added some minor comments. 

I think we can merge this very soon :-) 

To pass the `check_code_qualtiy` check, you can run `make style` from the root folder. 

Regarding the unit test, I think you can copy-paste a lot of the code from `test_roberta_modeling.py`. In order to skip tests for `torch_script, pruning, ..` (features we can add in another PR) you can do the same as is done in the reformer tests here: https://github.com/huggingface/transformers/blob/94cb73c2d2efeb188b522ff352f98b15124ba9f8/tests/test_modeling_reformer.py#L510 . 

If possible it would also be great to add some integration tests with your original model.

ibeltagy commented:
@bratao , @iliaschalkidis, about the custom CUDA kernel. We [recently](https://github.com/allenai/longformer/pull/27/files#diff-04c6e90faac2675aa89e2176d2eec7d8R4) added an efficient PyTorch implementation of our sliding window attention (without dilation), which is faster and uses less memory (with fp16) and should be easy to port to TF. The TVM code is still needed for the dilated sliding window attention, but dilation is more important for char-lm than fine-tuning on downstream tasks (QA, classification .. etc). This PR only uses the PyTorch code.

ibeltagy commented:
Thanks, @patrickvonplaten. I will address your comments and let you know. 

I also want to add `LongformerForSequenceClassification`, `LongformerForMultipleChoice`, `LongformerForTokenClassification`, `LongformerForQuestionAnswering` with automatic setting of global attention based on the task, so that the user doesn't need to worry about it. I will keep this to a different PR unless you think it is better to add it here.

sshleifer commented on src/transformers/modeling_longformer.py:
```diff
@@ -0,0 +1,667 @@
+
+class LongformerSelfAttention(nn.Module):
+    def __init__(self, config, layer_id):
+        super(LongformerSelfAttention, self).__init__()
```
just super().__init() I believe

sshleifer commented on src/transformers/modeling_longformer.py:
```diff
@@ -0,0 +1,667 @@
+        else:
+            raise ValueError(f'Expected values of `attention_mode` are "longformer" or "n2", given {config.attention_mode}')
+
+    def _pad_to_window_size(self, input_ids: torch.Tensor, attention_mask: torch.Tensor, token_type_ids: torch.Tensor,
```
do we use self? @staticmethod

sshleifer commented on src/transformers/tokenization_longformer.py:
```diff
@@ -0,0 +1,43 @@
+}
+
+
+class LongformerTokenizer(RobertaTokenizer):
```
Thx for reusing!

patrickvonplaten commented:
> Thanks, @patrickvonplaten. I will address your comments and let you know.
> 
> I also want to add `LongformerForSequenceClassification`, `LongformerForMultipleChoice`, `LongformerForTokenClassification`, `LongformerForQuestionAnswering` with automatic setting of global attention based on the task, so that the user doesn't need to worry about it. I will keep this to a different PR unless you think it is better to add it here.

It'd be great to add those as well :-) I think doing it in a second PR would be better as well

sshleifer commented on src/transformers/configuration_longformer.py:
```diff
@@ -0,0 +1,75 @@
+}
+
+
+class LongformerConfig(RobertaConfig):
```
Thanks for reusing!

sshleifer commented on src/transformers/modeling_longformer.py:
```diff
@@ -0,0 +1,667 @@
+# coding=utf-8
+# Copyright 2018 The Google AI Language Team Authors and The HuggingFace Inc. team.
```
typo I think

ibeltagy commented:
@patrickvonplaten, @sshleifer, I am done addressing your reviews. I still need to upload model weights but waiting to get access to the AllenAI organization account. Also, can you check the failed tests? them seem like issues with CircleCI rather than the code.

sshleifer commented on tests/test_modeling_longformer.py:
```diff
@@ -0,0 +1,246 @@
+
+import unittest
+
+import torch
```
this needs to be under `is_torch_available` to satisfy `CircleCI::run_tests_tf`

sshleifer commented on tests/test_modeling_longformer.py:
```diff
@@ -0,0 +1,246 @@
+
+    all_model_classes = (LongformerForMaskedLM, LongformerModel) if is_torch_available() else ()
+
+    class LongformerModelTester(object):
```
would you mind not indenting this and making it a module level object. I know it is done this way elsewhere, but we are trying to discontinue the practice.

sshleifer commented on tests/test_modeling_longformer.py:
```diff
@@ -0,0 +1,246 @@
+        self.model_tester.create_and_check_longformer_for_masked_lm(*config_and_inputs)
+
+    @slow
+    def test_model_from_pretrained(self):
```
I would delete this as it's covered my `test_inference_masked_lm`.

ibeltagy replied:
you mean delete `test_model_from_pretrained ` or `test_longformer_for_masked_lm` ?

ibeltagy replied:
ah, got it, it is already covered by  `test_inference_masked_lm` and `test_inference_no_head ` in `LongformerModelIntegrationTest `

sshleifer commented on tests/test_modeling_longformer.py:
```diff
@@ -0,0 +1,246 @@
+    @slow
+    def test_inference_masked_lm(self):
+        model = LongformerForMaskedLM.from_pretrained("longformer-base-4096")
+        input_ids = torch.tensor([[0] + [20920, 232, 328, 1437] * 1000 + [2]])  # long input
```
is this semantically meaningful?

ibeltagy commented:
@sshleifer, addressed your comments. Now `ci/circleci: run_tests_torch_and_tf` is timing out.

sshleifer commented:
I'm sorry.

Rebase may help the timeout, but I agree that it does not appear to be caused by this code.

I'd just verify that there aren't new tests that should be decorated `@slow` (append `--durations 0` to your local pytest command) and it should be fine.

ibeltagy commented:
The slowest one is 0.10s. 
```
0.10s call     tests/test_modeling_longformer.py::LongformerModelTest::test_save_load
0.10s call     tests/test_modeling_longformer.py::LongformerModelTest::test_attention_outputs
0.07s call     tests/test_modeling_longformer.py::LongformerModelTest::test_resize_tokens_embeddings
0.06s call     tests/test_modeling_longformer.py::LongformerModelTest::test_determinism
0.05s call     tests/test_modeling_longformer.py::LongformerModelTest::test_inputs_embeds
0.04s call     tests/test_modeling_longformer.py::LongformerModelTest::test_correct_missing_keys
0.04s call     tests/test_modeling_longformer.py::LongformerModelTest::test_hidden_states_output
0.04s call     tests/test_modeling_longformer.py::LongformerModelTest::test_longformer_model
0.03s call     tests/test_modeling_longformer.py::LongformerModelTest::test_initialization
0.03s call     tests/test_modeling_longformer.py::LongformerModelTest::test_longformer_for_masked_lm
0.02s call     tests/test_modeling_longformer.py::LongformerModelTest::test_model_common_attributes
```

Maybe the integration tests are slower than average because of the long document?

ibeltagy commented:
@patrickvonplaten, @sshleifer, looks like rebasing fixed the problem. Thanks.

AtrCy commented:
> ## Adding the Longformer model
> Paper: https://arxiv.org/abs/2004.05150
> The code is mostly copied from https://github.com/allenai/longformer
> Addressing issue: #3783
> 
> **TODOs**
> 
> * [x]  Add unit tests
> * [x]  Address PR reviews
> * [ ]  Upload model to HF instead of AllenAI s3 buckets.
> 
> **TODOs for other PRs**
> 
> * Add `LongformerForSequenceClassification`, `LongformerForMultipleChoice`, `LongformerForTokenClassification`, `LongformerForQuestionAnswering` with automatic setting of global attention based on the task.
> * Add `LongformerEncoderDecoder`
> * A few small TODOs in the code
> * Add an example to show how to convert existing pretrained models into their long version.

Is there any particular timeframe by when we can have LongformerForQuestionAnswering? Would love to use it in a small project of mine.
